controller/utils/badger.py

```diff
@@ -0,0 +1,125 @@
+  pass
+
+class BadgeValidator(pattern.EventEmitter):
+  """Class for badge authorization via HTTP API."""
+
+  def __init__(self, url, key_param, *args, **kwargs):
+    """Creates a BadgeValidator instance for validating badges.
+
+    The URL parameter will be appended with the customer key_param
+    as the parameter name and badge_id as the value when called.
+    ie:  ?myCustomKeyParam=<badge_id>
+
+    Args:
+      url: Authorization server URL
+      key_param: Name of badge parameter to add to the URL
+    """
+    super(BadgeValidator, self).__init__(*args, **kwargs)
+    self._url = url
+    self._key_param = key_param
+
+  def close(self):
+    """Closes the communication."""
+    super(BadgeReader, self).close()
+
+  def validate(self, badge_id):
+    """Validates the badge ID with a HTTP service. Any 2xx is considered success.
+
+    Args:
+      badge_id: String containing the badge identifier to authorize.
+    Returns:
+      Response boolean.
+    """
+    r = requests.get(url = self._url, params = {self._key_param: badge_id})
+    return r.status_code >= 200 and r.status_code < 300
+
+class BadgeReader(pattern.Worker, pattern.EventEmitter):
+  """Worker that monitors the badge reader and emits the badge ID on success.
+
+  This class supports Google's USB pcProx RFID Reader configured to support Google
+  employee badges.  However, it may support other USB HID based readers given
+  the proper vendor and product codes. It currently only supports reading the
+  0-9 and : ASCII characters.
+  """
+  def __init__(self, usb_vendor_id, usb_product_id,  *args, **kwargs):
+    """Creates a BadgeReader instance.
+
+    Args:
+      usb_vendor_id: USB Vendor ID of the device.
+      usb_product_id: USB Product ID of the device.
+    """
+    super(BadgeReader, self).__init__(worker_name='BadgeReader', *args, **kwargs)
+    self._device = None
+    self._last_badge_id = None
+    self._usb_vendor_id = usb_vendor_id
+    self._usb_product_id = usb_product_id
+    self._key_codes = {
+            2: u'1', 3: u'2', 4: u'3', 5: u'4', 6: u'5', 7: u'6',
+            8: u'7', 9: u'8', 10: u'9', 11: u'0', 39: u':'}
+
+  def _on_start(self):
+    devices = [InputDevice(path) for path in list_devices()]
+    for device in devices:
+      if device.info.vendor == int(self._usb_vendor_id, 16) and device.info.product == int(self._usb_product_id, 16):
+        self._device = device
+        self.logger.info("Badge reader device found.")
+        break
+    if self._device == None:
+      self.logger.warn("Device may be disconnected. Attempting to reconnect...")
+      time.sleep(10)
+      self._on_start()
+
+  def _on_run(self):
+    """Read and format an RFID badge ID
+
+    Raises:
+      BadgeReaderException: When device is unavailable, or reading fails.
+    Emits:
+      badge_read: A badge was successfully read.
+    """
+    try:
+      self._device.grab()
+      badge_id = ""
+      for event in self._device.read_loop():
```

Comment:
Is this loop going on forever?
If so, client may not be able to terminate properly. Try Ctrl+C the controller client and see if it exits by itself.